src/arithmetic/algebraic_expression.c

```diff
@@ -232,8 +231,15 @@ static void _RemovePathFromGraph(QueryGraph *g, QGEdge **path) {
 		QGNode *dest = e->dest;
 
 		QueryGraph_RemoveEdge(g, e);
-		if(QGNode_EdgeCount(src) == 0) QueryGraph_RemoveNode(g, src);
-		if(QGNode_EdgeCount(dest) == 0) QueryGraph_RemoveNode(g, dest);
+		QGEdge_Free(e);
+		if(QGNode_EdgeCount(src) == 0) {
+			QueryGraph_RemoveNode(g, src);
+			QGNode_Free(src);
```

Comment:
Freeing either the src or dest node at this point is dangerous as the algebraic expression constructed for the path have references to the first and last node on the path. in some cases where the edge is also referenced the expression will hold a pointer to it.
Maybe if we move from algebraic expressions holding references to QueryGraph objects (nodes/edges) to simply aliases this entire problem would be solved.

Reply:
Not dangerous because this call directly follows `AlgebraicExpression_UpdateReferences`, so that the expressions we actually keep hold pointers to the real QueryGraph rather than to the `path` here (which is derived from the cloned QueryGraph).

This duality is way too complicated, in my opinion. Switching to aliases might be a valid fix! We do need some QG entity data at other points (like collecting relation types in OpConditionalTraverse), but we should be able to retrieve the QG entity by alias at that time. I'll look into it!